src/eval/metric_completion.py

```diff
@@ -0,0 +1,196 @@
+
+        games_states, games_actions, games_target_returns, games_seq_len = [], [], [], []
+        # for seq in sequences:
+
+        if self.single_player:
+            states, actions, target_returns, seq_len = self.single_player_prep(sequences[5])
+        else:
+            states, actions, target_returns, seq_len = self.two_player_prep(sequences[5])
+
+        games_states.append(states)
+        games_actions.append(actions)
+        games_target_returns.append(target_returns)
+        games_seq_len.append(seq_len)
+
+        return games_states, games_actions, games_target_returns, games_seq_len
+
+    def single_player_prep(self, seq):
+        pass
+
+    def two_player_prep(self, seq):
+        move_indices, board_tensors, end_rewards, boards = translate.encode_seq(seq, self.board_to_tensor, self.shaping)
+        seq_len = len(board_tensors)
+
+        action_seq = torch.nn.functional.one_hot(torch.tensor(move_indices, dtype=int), num_classes=self.action_dim)
+        actions = action_seq.reshape(1, seq_len, self.action_dim).to(device=self.device, dtype=torch.float32)
+
+        state_seq = torch.stack(board_tensors)
+        states = state_seq.reshape(1, seq_len, self.state_dim).to(device=self.device, dtype=torch.float32)
+
+        black_seq_len = seq_len // 2
+        white_seq_len = seq_len - black_seq_len
+
+        if self.shaping:
+            white_evals = []
+            black_evals = []
+            for b_idx, b in enumerate(boards):
+                info = self.engine.analyse(
+                    b, chess.engine.Limit(depth=self.shaping_eval_depth), info=chess.engine.Info.SCORE
+                )
+                if b_idx % 2 == 0:
+                    white_evals.append(info["score"].white().score(mate_score=10000) / 10000)
+                else:
+                    black_evals.append(info["score"].black().score(mate_score=10000) / 10000)
+            white_evals = torch.tensor(white_evals, dtype=torch.float32, device=self.device)
+            black_evals = torch.tensor(black_evals, dtype=torch.float32, device=self.device)
+            white_result = torch.full((len(white_evals),), end_rewards[0], device=self.device)
+            black_result = torch.full((len(black_evals),), end_rewards[1], device=self.device)
+
+            white_returns = torch.sub(white_result, white_evals)
+            black_returns = torch.sub(black_result, black_evals)
+
+        else:
+            black_returns = self.discount ** torch.arange(black_seq_len, device=self.device) * end_rewards[1]
+            white_returns = self.discount ** torch.arange(white_seq_len, device=self.device) * end_rewards[0]
+
+        condition = torch.arange(seq_len, device=self.device) % 2 == 0
+        target_returns = torch.zeros(1, seq_len, 1, device=self.device, dtype=torch.float32)
+        target_returns[:, condition, :] = white_returns.reshape(1, white_seq_len, 1)
+        target_returns[:, ~condition, :] = black_returns.reshape(1, black_seq_len, 1)
```

Comment:
Return to go are actually max_return - returns

Reply:
i think this is resolved now because all of the work with return to go are now in the dataset/translate module. Let me know if you think what I did in those files with return to go is correct